pkg/cri/resource-manager/control/cpu/cpu.go

```diff
@@ -113,36 +127,117 @@ func (ctl *cpuctl) PostStopHook(c cache.Container) error {
 	return nil
 }
 
-// enforce enforces a class-specific configuration to a cpuset
-func (ctl *cpuctl) enforce(class string, cpus ...int) error {
+// enforceCpufreq enforces a class-specific cpufreq configuration to a cpuset
+func (ctl *cpuctl) enforceCpufreq(class string, cpus ...int) error {
 	if _, ok := ctl.config.Classes[class]; !ok {
 		return fmt.Errorf("non-existent cpu class %q", class)
 	}
 
-	log.Debug("enforcing cpu class %q on %v", class, cpus)
+	min := int(ctl.config.Classes[class].MinFreq)
+	max := int(ctl.config.Classes[class].MaxFreq)
+	log.Debug("enforcing cpu frequency limits {%d, %d} from class %q on %v", min, max, class, cpus)
 
-	if err := utils.SetCPUsScalingMinFreq(cpus, (int)(ctl.config.Classes[class].MinFreq)); err != nil {
-		return fmt.Errorf("Cannot set min freq %d: %w", ctl.config.Classes[class].MinFreq, err)
+	if err := utils.SetCPUsScalingMinFreq(cpus, min); err != nil {
+		return fmt.Errorf("Cannot set min freq %d: %w", min, err)
 	}
 
-	if err := utils.SetCPUsScalingMaxFreq(cpus, (int)(ctl.config.Classes[class].MaxFreq)); err != nil {
-		return fmt.Errorf("Cannot set max freq %d: %w", ctl.config.Classes[class].MaxFreq, err)
+	if err := utils.SetCPUsScalingMaxFreq(cpus, max); err != nil {
+		return fmt.Errorf("Cannot set max freq %d: %w", max, err)
 	}
 
 	return nil
 }
 
+// enforceUncore enforces uncore frequency limits
+func (ctl *cpuctl) enforceUncore(assignments cpuClassAssignments, affectedCPUs ...int) error {
+	cpus := cpuset.NewCPUSet(affectedCPUs...)
+
+	for _, cpuPkgID := range ctl.system.PackageIDs() {
+		cpuPkg := ctl.system.Package(cpuPkgID)
+		for _, cpuDieID := range cpuPkg.DieIDs() {
+			dieCPUs := cpuPkg.DieCPUSet(cpuDieID)
+
+			// Check if this die is affected by the specified cpuset
+			if cpus.Size() == 0 || dieCPUs.Intersection(cpus).Size() > 0 {
+				min, max, minCls, maxCls := effectiveUncoreFreqs(utils.NewIDSet(dieCPUs.ToSlice()...), ctl.config.Classes, assignments)
+
+				if min == 0 && max == 0 {
+					log.Debug("no uncore frequency limits for cpu package/die %d/%d", cpuPkgID, cpuDieID)
+					continue
+				} else if !utils.UncoreFreqAvailable() {
+					return fmt.Errorf("unable to set uncore frequency limits, driver not available, make sure that the intel_uncore_frequency driver is loaded")
+				}
+
+				log.Debug("enforcing uncore min freq to %d (class %q), max freq to %d (class %q) on cpu package/die %d/%d", min, minCls, max, maxCls, cpuPkgID, cpuDieID)
+				if min > 0 {
```

Comment:
Is it possible that user is able to set min > max

Reply:
Yes it is. I added a warning log message if this is detected.